### PR TITLE
Move loading of .nodes and .edges to io.hpp

### PR DESCRIPTION
**Pull request by chaupow**
# Issue

Working on data loading refactoring captured here: 
https://github.com/Project-OSRM/osrm-backend/issues/2989

This issue is working on `.nodes` and `.edges` files

## Tasklist
 - [ ]  wait for comment on the TODO in `io.hpp`
 - [x] wait for https://github.com/Project-OSRM/osrm-backend/pull/3074 to be merged
 - [x] wait for https://github.com/Project-OSRM/osrm-backend/pull/3090 to be merged
 - [ ] review
 - [ ] adjust for comments

/cc @karenzshea @TheMarex  @miccolis @danpat

### CODE REVIEW

File: include/engine/datafacade/internal_datafacade.hpp [hunk 1]
```diff
@@ -174,46 +177,42 @@ class InternalDataFacade final : public BaseDataFacade
...
-        boost::filesystem::ifstream edges_input_stream(edges_file, std::ios::binary);
-        unsigned number_of_edges = 0;
-        edges_input_stream.read((char *)&number_of_edges, sizeof(unsigned));
+        auto number_of_edges = storage::io::readEdgesSize(edges_input_stream);
         m_via_geometry_list.resize(number_of_edges);
         m_name_ID_list.resize(number_of_edges);
         m_turn_instruction_list.resize(number_of_edges);
         m_lane_data_id.resize(number_of_edges);
         m_travel_mode_list.resize(number_of_edges);
         m_entry_class_id_list.resize(number_of_edges);
```
Comment (MoKob) [inline include/engine/datafacade/internal_datafacade.hpp]: just a heads up, I am just about to merge the turn-angle branch which adds to this list.

File: include/storage/io.hpp [hunk 2]
```diff
@@ -61,6 +62,88 @@ void readHSGR(boost::filesystem::ifstream &input_stream,
...
+void readNodesData(boost::filesystem::ifstream &nodes_input_stream,
+                   CoordinateT *coordinate_list,
+                   OSMNodeIDVectorT &osmnodeid_list,
+                   std::uint32_t number_of_coordinates)
+{
+    extractor::QueryNode current_node;
+    for (std::uint32_t i = 0; i < number_of_coordinates; ++i)
+    {
+        nodes_input_stream.read((char *)&current_node, sizeof(extractor::QueryNode));
+        coordinate_list[i] = util::Coordinate(current_node.lon, current_node.lat);
```
Comment (MoKob) [inline include/storage/io.hpp]: `util:Coordinate` vs `CoordinateT` mismatch. Shouldn't `{}` work?

File: include/engine/datafacade/internal_datafacade.hpp (hunk 1 above)
Comment (chaupow) [inline include/engine/datafacade/internal_datafacade.hpp]: sorry, but what does that mean?

Comment (chaupow) [inline include/engine/datafacade/internal_datafacade.hpp]: ah ok yea I think I can manage that :D thx

File: include/storage/io.hpp (hunk 2 above)
Comment (chaupow) [inline include/storage/io.hpp]: fixed thanks :D

File: include/storage/io.hpp [hunk 3]
```diff
@@ -61,6 +62,88 @@ void readHSGR(boost::filesystem::ifstream &input_stream,
...
+    nodes_input_stream.read((char *)&number_of_coordinates, sizeof(std::uint32_t));
+    return number_of_coordinates;
+}
+
+// Reads coordinates and OSM node IDs from .nodes files
+// TODO: OSMNodeIDVectorT kind of breaks the consistency of the input parameters
+// Any ideas how to fix this?
+template <typename CoordinateT, typename OSMNodeIDVectorT>
+void readNodesData(boost::filesystem::ifstream &nodes_input_stream,
+                   CoordinateT *coordinate_list,
```
Comment (chaupow) [inline include/storage/io.hpp]: `storage.cpp` is using this function to read the nodes data and store it in shared memory e.g. with
```
util::Coordinate *coordinates_ptr = shared_layout_ptr->GetBlockPtr<util::Coordinate, true>(shared_memory_ptr, SharedDataLayout::COORDINATE_LIST);
```
so I am not sure how to do this without plain pointers? Are there ways?

Comment (chaupow) [inline include/storage/io.hpp]: @MoKob can you check whether the new commit does what you mean? if not I need a little bit of clarification please :)

File: include/storage/io.hpp [hunk 4]
```diff
@@ -63,20 +65,104 @@ void readHSGR(boost::filesystem::ifstream &input_stream,
...
+    //     edges_input_stream.read((char *)&(current_edge_data), sizeof(extractor::OriginalEdgeData));
+    //     via_geometry_ptr[i] = current_edge_data.via_geometry;
+    //     name_id_ptr[i] = current_edge_data.name_id;
+    //     travel_mode_ptr[i] = current_edge_data.travel_mode;
+    //     lane_data_id_ptr[i] = current_edge_data.lane_data_id;
+    //     turn_instructions_ptr[i] = current_edge_data.turn_instruction;
+    //     entry_class_id_ptr[i] = current_edge_data.entry_classid;
+    //     pre_turn_bearing_ptr[i] = current_edge_data.pre_turn_bearing;
+    //     post_turn_bearing_ptr[i] = current_edge_data.post_turn_bearing;
+    // }
```
Comment (chaupow) [inline include/storage/io.hpp]: forgot to delete these lines 
TODO

**Comment by chaupow** [comment]
@MoKob can you check whether the adjustments are ok now?

File: include/storage/io.hpp [hunk 5]
```diff
@@ -51,9 +53,9 @@ inline HSGRHeader readHSGRHeader(boost::filesystem::ifstream &input_stream)
 // Needs to be called after HSGRHeader() to get the correct offset in the stream
 template <typename NodeT, typename EdgeT>
 void readHSGR(boost::filesystem::ifstream &input_stream,
-              NodeT *node_buffer,
+              NodeT node_buffer[],
               std::uint32_t number_of_nodes,
-              EdgeT *edge_buffer,
+              EdgeT edge_buffer[],
```
Comment (daniel-j-h) [inline include/storage/io.hpp]: These can still be nullptr

File: include/storage/io.hpp [hunk 6]
```diff
@@ -63,20 +65,88 @@ void readHSGR(boost::filesystem::ifstream &input_stream,
...
+{
+    std::uint32_t number_of_coordinates;
+    nodes_input_stream.read((char *)&number_of_coordinates, sizeof(std::uint32_t));
+    return number_of_coordinates;
+}
+
+// Reads coordinates and OSM node IDs from .nodes files
+template <typename CoordinateT, typename OSMNodeIDVectorT>
+void readNodesData(boost::filesystem::ifstream &nodes_input_stream,
+                   CoordinateT coordinate_list[],
```
Comment (daniel-j-h) [inline include/storage/io.hpp]: and here

File: include/storage/io.hpp [hunk 7]
```diff
@@ -63,20 +65,88 @@ void readHSGR(boost::filesystem::ifstream &input_stream,
...
+                   OSMNodeIDVectorT &osmnodeid_list,
+                   std::uint32_t number_of_coordinates)
+{
+    extractor::QueryNode current_node;
+    for (std::uint32_t i = 0; i < number_of_coordinates; ++i)
+    {
+        nodes_input_stream.read((char *)&current_node, sizeof(extractor::QueryNode));
+        coordinate_list[i] = CoordinateT(current_node.lon, current_node.lat);
+        osmnodeid_list.push_back(current_node.node_id);
```
Comment (daniel-j-h) [inline include/storage/io.hpp]: .reserve the vector upfront - you know its final size

File: include/storage/io.hpp (hunk 5 above)
Comment (chaupow) [inline include/storage/io.hpp]: now boost_asserted

File: include/storage/io.hpp [hunk 8]
```diff
@@ -63,20 +65,88 @@ void readHSGR(boost::filesystem::ifstream &input_stream,
...
+          typename PostTurnBearingT>
+void readEdgesData(boost::filesystem::ifstream &edges_input_stream,
+                   GeometryIDT geometry_list[],
+                   NameIDT name_id_list[],
+                   TurnInstructionT turn_instruction_list[],
+                   LaneDataIDT lane_data_id_list[],
+                   TravelModeT travel_mode_list[],
+                   EntryClassIDT entry_class_id_list[],
+                   PreTurnBearingT pre_turn_bearing_list[],
+                   PostTurnBearingT post_turn_bearing_list[],
```
Comment (chaupow) [inline include/storage/io.hpp]: now boost_asserted

File: include/storage/io.hpp (hunk 6 above)
Comment (chaupow) [inline include/storage/io.hpp]: now boost_asserted

File: include/storage/io.hpp [hunk 9]
```diff
@@ -63,20 +65,88 @@ void readHSGR(boost::filesystem::ifstream &input_stream,
...
+    nodes_input_stream.read((char *)&number_of_coordinates, sizeof(std::uint32_t));
+    return number_of_coordinates;
+}
+
+// Reads coordinates and OSM node IDs from .nodes files
+template <typename CoordinateT, typename OSMNodeIDVectorT>
+void readNodesData(boost::filesystem::ifstream &nodes_input_stream,
+                   CoordinateT coordinate_list[],
+                   OSMNodeIDVectorT &osmnodeid_list,
+                   std::uint32_t number_of_coordinates)
```
Comment (chaupow) [inline include/storage/io.hpp]: Compared to the other lists, osmnodeid_list is stored in a packed vector.

**Comment by karenzshea** [comment]
Closing in favor of https://github.com/Project-OSRM/osrm-backend/pull/3148

**Pull request closed without merging**